### PR TITLE
[Core][ModelPart] SetBufferSize not working with SubSubModelpar…

### LINKED ISSUE #5129: [ModelPart] SetBufferSize not working with SubSubModelpart
The `SetBufferSize` method does not work when SubSubModelParts exist because it is not (because it cannot) called recursively, but only set directly

https://github.com/KratosMultiphysics/Kratos/blob/e525bc89238005dc01b7f2c0b3d0a61995e4201d/kratos/sources/model_part.cpp#L1563-L1585

I guess we need a (private) Method e.g. `SetBufferSizeSubModelParts` that can be called recursively

### PR DESCRIPTION
**Pull request by loumalouomega**
Fixes #5129

Additionally I added some documentation. I also imported properly Kratos on the test file

Test added

### CODE REVIEW

**Comment by loumalouomega** [comment]
@pooyan-dadvand which is the point of mIndices? This looks unused, and only appears on the constructor

**Comment by loumalouomega** [comment]
@philbucher

File: kratos/sources/model_part.cpp [hunk 1]
```diff
@@ -1584,6 +1583,15 @@ void ModelPart::SetBufferSize(ModelPart::IndexType NewBufferSize)
 
 }
 
+void ModelPart::SetBufferSizeSubModelParts(ModelPart::IndexType NewBufferSize)
+{
+    for(auto it_sub_model_part = mSubModelParts.begin(); it_sub_model_part != mSubModelParts.end(); ++it_sub_model_part) {
```
Comment (philbucher) [inline kratos/sources/model_part.cpp]:
```suggestion
    for(auto& r_sub_model_part : mSubModelParts) {
```

File: kratos/sources/model_part.cpp [hunk 2]
```diff
@@ -1584,6 +1583,15 @@ void ModelPart::SetBufferSize(ModelPart::IndexType NewBufferSize)
...
+        it_sub_model_part->SetBufferSizeSubModelParts(NewBufferSize);
```
Comment (philbucher) [inline kratos/sources/model_part.cpp]:
```suggestion
        r_sub_model_part.SetBufferSizeSubModelParts(NewBufferSize);
```

File: kratos/sources/model_part.cpp [hunk 3]
```diff
@@ -1566,9 +1566,8 @@ void ModelPart::SetBufferSize(ModelPart::IndexType NewBufferSize)
         << Name() << " please call the one of the root model part: "
         << GetRootModelPart().Name() << std::endl;
 
-    for(SubModelPartIterator i_sub_model_part = mSubModelParts.begin(); i_sub_model_part != mSubModelParts.end(); i_sub_model_part++)
-    {
-        i_sub_model_part->mBufferSize = NewBufferSize;
+    for(auto it_sub_model_part = mSubModelParts.begin(); it_sub_model_part != mSubModelParts.end(); ++it_sub_model_part) {
```
Comment (philbucher) [inline kratos/sources/model_part.cpp]: same

**Review by philbucher — COMMENTED** [review]
minor comment otherwise looks good to me, thx

**Comment by loumalouomega** [comment]
Done @philbucher

File: kratos/includes/model_part.h [hunk 4]
```diff
@@ -1532,27 +1545,27 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
...
+    IndexType mBufferSize; /// The buffers size of the database
 
-    ProcessInfo::Pointer mpProcessInfo;
+    ProcessInfo::Pointer mpProcessInfo; /// The process info instance
 
-    TablesContainerType mTables;
+    TablesContainerType mTables; /// The tables contained on the model part
 
-    std::vector<IndexType> mIndices;
+    std::vector<IndexType> mIndices; /// The list of indices /// NOTE: This looks unused, only appears on the constructor @pooyan_dadvand
```
Comment (philbucher) [inline kratos/includes/model_part.h:1556]: ping @pooyan-dadvand

**Review by philbucher — APPROVED** [review]
cool, thx

File: kratos/sources/model_part.cpp [hunk 5]
```diff
@@ -1584,6 +1582,15 @@ void ModelPart::SetBufferSize(ModelPart::IndexType NewBufferSize)
 
 }
 
+void ModelPart::SetBufferSizeSubModelParts(ModelPart::IndexType NewBufferSize)
```
Comment (RiccardoRossi) [inline kratos/sources/model_part.cpp:1585]: guys, this is NOT correct.

you can change the buffer size only of the root modelpart, and the behaviour is deliberate

current implementation is the following:

~~~cpp
void ModelPart::SetBufferSize(ModelPart::IndexType NewBufferSize)
{
    KRATOS_ERROR_IF(IsSubModelPart()) << "Calling the method of the sub model part "
        << Name() << " please call the one of the root model part: "
        << GetRootModelPart().Name() << std::endl;

    for(SubModelPartIterator i_sub_model_part = mSubModelParts.begin(); i_sub_model_part != mSubModelParts.end(); i_sub_model_part++)
    {
        i_sub_model_part->mBufferSize = NewBufferSize;
    }

    mBufferSize = NewBufferSize;

    auto nodes_begin = NodesBegin();
    const int nnodes = static_cast<int>(Nodes().size());
    #pragma omp parallel for firstprivate(nodes_begin,nnodes)
    for(int i = 0; i<nnodes; ++i)
    {
        auto node_iterator = nodes_begin + i;
        node_iterator->SetBufferSize(mBufferSize);
    }

}
~~~

**Review by RiccardoRossi — COMMENTED** [review]
ok now i understdood. forget my last comment

File: kratos/includes/model_part.h (hunk 4 above)
Comment (pooyan-dadvand) [reply kratos/includes/model_part.h:1556]: Historical residuals. AKA to be removed....

**Comment by pooyan-dadvand** [comment]
> @pooyan-dadvand which is the point of mIndices? This looks unused, and only appears on the constructor

Saw your comment too late. Can be safely deleted

**Pull request merged**